### PR TITLE
raises error when cf push failed

### LINKED ISSUE #441: Cloud Foundry Deployment exits with status code 0 even if cf push fails
If the `cf push` step fails, the exit code should be 1. Currently, this is causing my builds to be marked a success, even if the application fails to deploy.

### PR DESCRIPTION
**Pull request by shawnzhu**
fixes #441

### CODE REVIEW

**Comment by jossy** [comment]
Getting this merged would be super helpful! I can't see how Travis' Cloud Foundry deployment can be used for non-trivial projects if a build does not fail on push error.

**Comment by afeld** [comment]
Hmm, seems that `context.shell` won't throw an error if a command fails, meaning that there would need to be a corresponding change everywhere in the codebase `context.shell` is invoked. Thinking it would be much simpler to have `context.shell` throw an error when the underlying `system` call returns `false`, rather than putting the onus on the callers to check for it.

**Comment by umaar** [comment]
Right now, a if a Cloud Foundry deployment fails, even it returns with a failing exit code, this is not picked up by Travis and as such, the Travis build is marked as successful even the deployment failed. Could this PR be merged?

File: lib/dpl/provider/cloud_foundry.rb [hunk 1]
```diff
@@ -23,8 +23,8 @@ def needs_key?
       end
 
       def push_app
-        context.shell "./cf push#{manifest}"
-        context.shell "./cf logout"
+        error 'Failed to push app' if context.shell("./cf push#{manifest}") == false
```
Comment (BanzaiMan) [inline lib/dpl/provider/cloud_foundry.rb]: `if x == false` is cumbersome.

```ruby
error 'Failed to push app' unless context.shell("./cf push#{manifest}")
```

File: lib/dpl/provider/cloud_foundry.rb [hunk 2]
```diff
@@ -23,8 +23,8 @@ def needs_key?
...
+        ensure context.shell "./cf logout"
```
Comment (BanzaiMan) [inline lib/dpl/provider/cloud_foundry.rb]: `ensure` block should be introduced here.

```ruby
⋮
ensure
  context.shell "./cf logout"
⋮
```

File: lib/dpl/provider/cloud_foundry.rb [hunk 3]
```diff
@@ -23,8 +23,10 @@ def needs_key?
       end
 
       def push_app
-        context.shell "./cf push#{manifest}"
-        context.shell "./cf logout"
+        error 'Failed to push app' unless context.shell("./cf push#{manifest}")
```
Comment (BanzaiMan) [inline lib/dpl/provider/cloud_foundry.rb]: This `context.shell` call appears to return `false` in specs. Could you look into this?

File: lib/dpl/provider/cloud_foundry.rb [hunk 4]
```diff
@@ -23,8 +23,10 @@ def needs_key?
...
+
+        ensure
```
Comment (BanzaiMan) [inline lib/dpl/provider/cloud_foundry.rb]: Stylistically, this line and the line below should be one level up (delete 2 spaces at the beginning of the lines).

**Comment by shawnzhu** [comment]
@BanzaiMan review comments addressed, would you please review it again?

**Comment by BanzaiMan** [comment]
@shawnzhu Great! Do you have example builds that show how this would work (one for success, one for failure)?

**Comment by shawnzhu** [comment]
@BanzaiMan no. I need to figure out how to show outcome of this change in travis-ci.org :)

what is the special configuration in `.travis.yml` to use my branch of `travis-ci/dpl`?

**Comment by BanzaiMan** [comment]
@shawnzhu https://github.com/travis-ci/dpl/blob/master/TESTING.md

**Comment by shawnzhu** [comment]
Here's one example:
* source: https://github.com/shawnzhu/get-started-python/tree/test-dpl
* [`.travis.yml`](https://github.com/shawnzhu/get-started-python/blob/test-dpl/.travis.yml)
* a failed deployment fails Travis build: https://travis-ci.org/shawnzhu/get-started-python/builds/230193532

**Comment by jwickens** [comment]
Would be great to get this merged!

**Pull request merged**